runtime/service.go

```diff
@@ -1149,54 +1135,70 @@ func (s *service) Connect(requestCtx context.Context, req *taskAPI.ConnectReques
-// Shutdown is not directly exposed to containerd clients.
+// containerd calls this API on behalf of the user in the following cases:
+// * After any task is deleted via containerd's API
+// * After any task Create call returns an error
 func (s *service) Shutdown(requestCtx context.Context, req *taskAPI.ShutdownRequest) (*ptypes.Empty, error) {
 	defer logPanicAndDie(log.G(requestCtx))
-	s.logger.WithFields(logrus.Fields{"task_id": req.ID, "now": req.Now}).Debug("shutdown")
+	s.logger.WithFields(logrus.Fields{"task_id": req.ID, "now": req.Now}).Debug("Shutdown")
 
 	shouldShutdown := req.Now || s.exitAfterAllTasksDeleted && s.taskManager.ShutdownIfEmpty()
 	if !shouldShutdown {
 		return &ptypes.Empty{}, nil
 	}
 
-	// cancel the shim context no matter what, which will result in the VM getting a SIGKILL (if not already
-	// dead from graceful shutdown) and the shim process itself to begin exiting
-	defer s.shimCancel()
-
 	s.logger.Info("stopping the VM")
 
-	var shutdownErr error
+	shutdownCtx, shutdownCancel := context.WithTimeout(context.Background(), defaultShutdownTimeout*time.Second)
+	defer shutdownCancel()
 
-	_, err := s.agentClient.Shutdown(requestCtx, req)
-	if err != nil {
-		shutdownErr = multierror.Append(shutdownErr, errors.Wrap(err, "failed to shutdown VM Agent"))
+	if err := s.shutdown(requestCtx, shutdownCtx, req); err != nil {
+		return &ptypes.Empty{}, err
 	}
 
-	err = s.machine.StopVMM()
-	if err != nil {
-		shutdownErr = multierror.Append(shutdownErr, errors.Wrap(err, "failed to gracefully stop VM"))
+	return &ptypes.Empty{}, nil
+}
+
+// shutdown will shutdown the VMM. requestCtx is used to issue requests and shutdownCtx is used for waiting a graceful shutdown procedure.
+// When shutdownCtx's Done channel is closed during the graceful shutdown procedure, the VMM is forcefully killed.
+func (s *service) shutdown(requestCtx, shutdownCtx context.Context, req *taskAPI.ShutdownRequest) error {
+	if _, err := s.agentClient.Shutdown(requestCtx, req); err != nil {
+		return err
 	}
 
-	err = os.RemoveAll(s.shimDir.RootPath())
-	if err != nil {
-		shutdownErr = multierror.Append(shutdownErr, errors.Wrapf(err, "failed to remove VM dir %q during shutdown", s.shimDir.RootPath()))
+	// cancel the shim at the end, in case even the last StopVMM() doesn't work
+	defer s.shimCancel()
+
+	shutdownCh := make(chan error)
+	go func() {
+		defer close(shutdownCh)
+		err := s.machine.Wait(shutdownCtx)
+		shutdownCh <- err
+	}()
+
+	var waitErr error
+	select {
+	case err := <-shutdownCh: // err = nil if the VMM exists successfully
+		waitErr = err
+	case <-shutdownCtx.Done():
+		waitErr = status.Error(codes.DeadlineExceeded, "timed out waiting for VM shutdown")
 	}
 
-	if shutdownErr != nil {
-		s.logger.WithError(shutdownErr).Error()
-		return nil, shutdownErr
+	if waitErr != nil {
+		if err := s.machine.StopVMM(); err != nil {
+			return multierror.Append(waitErr, err)
+		}
```

Comment:
I'm wondering if we need to call `StopVMM` at all. AFAIK (correct me if wrong), Firecracker wouldn't have any different behavior between a SIGTERM and a SIGKILL because they don't have a special SIGTERM handler, so I think this should just have the same end effect as cancelling the context (which results in SIGKILL).

If we do want to leave it in, then I think it should not be done when `waitErr` indicates that the machine exited with non-zero code (as the machine is already dead, so no need to SIGTERM it).

Reply:
I wanted to capture an error from StopVMM() by explicitly calling the function. It doesn't have much right now though. https://github.com/firecracker-microvm/firecracker-go-sdk/blob/43d336c3dcbf623d7dec9241f88ca9c0afb44e11/machine.go#L521

Reply:
I think cancelling the context to kill Firecracker depends too much of the current implementation: 1) Firecracker doesn't distinguish SIGTERM/SIGKILL and 2) StopVMM is just SIGTERM. So the last revision still has an explicit StopVMM call.